Java: model top 100 JDK APIs

jcogs33 opened the pull request:
### Description
This PR adds MaD models for top 100 JDK APIs which were not previously modeled. It also adds a test case that checks that a manual model exists for each of the top 100 APIs.

### Consideration
**Models:**
- Please take a close look at the models added for exceptions as I'm not sure if I handled the field flow for them correctly.
- I left the following two apis unmodeled based on comments [here](https://github.com/github/codeql/pull/11572#discussion_r1040978328), and [here](https://github.com/github/codeql/pull/11572#discussion_r1040991878). Let me know if I should add a model for either of these after all: `java.lang.String#valueOf(Object)` and `java.lang.Throwable#printStackTrace()`.
- I put `java.lang.String#charAt(int)` as a positive model based on the comment [here](https://github.com/github/codeql/pull/11572#discussion_r1040980681), let me know if I should remove the model instead.

**Test Case:**
- Let me know if the test case needs to be changed. I'm not sure if I structured it as intended.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.util.stream", "Stream", True, "map", "(Function)", "", "Argument[-1].Element", "Argument[0].Parameter[0]", "value", "manual"]     # SUPPORTED: already modelled in java.util.stream.model.yml at line 45
+      - ["java.util.stream", "Stream", True, "map", "(Function)", "", "Argument[0].ReturnValue", "ReturnValue.Element", "value", "manual"]       # SUPPORTED: already modelled in java.util.stream.model.yml at line 46
+      - ["java.util.stream", "Stream", True, "forEach", "(Consumer)", "", "Argument[-1].Element", "Argument[0].Parameter[0]", "value", "manual"] # SUPPORTED (extra one from Michael's list): already modelled in java.util.stream.model.yml at line 33
+      # - ["java.util.stream", "Stream", True, "collect", "(Collector)", "", "", "", "value", "manual"]                                          # **UNSUPPORTED**: noted in java.util.stream.model.yml at line 19 that "collect(Collector<T,A,R> collector) is handled separately on a case-by-case basis as it is too complex for MaD" (Note: look into if/how this is ever handled)
```
Yes, it's handled in https://github.com/github/codeql/blob/main/java/ql/lib/semmle/code/java/frameworks/Stream.qll - at least common cases of it is - it needs per-collector special-casing.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.lang", "Integer", False, "parseInt", "(String)", "", "Argument[0]", "ReturnValue", "taint", "manual"]                             # SUPPORTED: no MaD row or regular CodeQL model that I can find... (Note: check for existing model again)
+      - ["java.io", "File", False, "File", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]                                     # SUPPORTED: already modelled in java.io.model.yml at line 60, but no signature specified in existing model
+      - ["java.io", "PrintWriter", False, "write", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]                             # SUPPORTED: no MaD row, modelled by Writer.write in java.io.model.yml at line 86 through subtyping
+      - ["java.nio.file", "Path", True, "resolve", "(String)", "", "Argument[0]", "ReturnValue", "taint", "manual"]                              # **COLLISION** (both supported and unsupported per initial telemetry query): already modelled in java.nio.file.model.yml at line 29, but no signature specified in existing model
```
🤔 "both supported and unsupported" that sounds weird. @atorralba do you know what's up?

jcogs33 replied:
Sorry, I should have removed that comment. I discussed the "collision" cases with @michaelnebel earlier.
From Michael:
> ...it looks like the model `java.nio.file;Path;true;resolve;;;Argument[-1..0];ReturnValue;taint;manual` was added on August 29th this year. From when it is merged into the source code until it is released and in production can take some time. AFAIK the telemetry stored in the datawarehouse is stored for around 3 months. This means that during the last three months there has been points in time, where the API was supported and points in time where the API was not supported and I think that is the explanation of the difference.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.lang", "IllegalArgumentException", False, "IllegalArgumentException", "(String)", "", "Argument[0]", "Argument[-1]", "value", "manual"]           # **UNSUPPORTED**: flow from arg to new IllegalArgumentException object?
+      - ["java.lang", "IllegalStateException", False, "IllegalStateException", "(String)", "", "Argument[0]", "Argument[-1]", "value", "manual"]                 # **UNSUPPORTED**: flow from arg to new IllegalStateException object?
+      - ["java.lang", "UnsupportedOperationException", False, "UnsupportedOperationException", "(String)", "", "Argument[0]", "Argument[-1]", "value", "manual"] # **UNSUPPORTED**: flow from arg to new UnsupportedOperationException object?
+      - ["java.util.stream", "Collectors", False, "toList", "()", "", "Argument[-1].Element", "ReturnValue.Element", "value", "manual"]                          # **UNSUPPORTED**: seems to have flow from input elements into a Collector->List?, is this too complex for MaD like Stream.collect(Collector)?
```
This shouldn't have a model - for all intents and purposes it's effectively an enum value to be used as input to Stream.collect. And it's handled there.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.io", "File", False, "File", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]                                     # SUPPORTED: already modelled in java.io.model.yml at line 60, but no signature specified in existing model
+      - ["java.io", "PrintWriter", False, "write", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]                             # SUPPORTED: no MaD row, modelled by Writer.write in java.io.model.yml at line 86 through subtyping
+      - ["java.nio.file", "Path", True, "resolve", "(String)", "", "Argument[0]", "ReturnValue", "taint", "manual"]                              # **COLLISION** (both supported and unsupported per initial telemetry query): already modelled in java.nio.file.model.yml at line 29, but no signature specified in existing model
+      - ["java.math", "BigDecimal", False, "BigDecimal", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]                       # SUPPORTED: no MaD row or regular CodeQL model that I can find... (Note: check for existing model again)
```
Don't think this should have a model - we're mostly not concerned about numbers. They're usually not taint-carriers.

jcogs33 replied:
Okay thanks, that's what I thought at first, but then got confused by the fact that it was supposedly "supported".
I'll take a look at it again to see why it came up as supported in the telemetry query.

atorralba replied:
As mentioned below, I'd prefer having this kind of models.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.math", "BigDecimal", False, "BigDecimal", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]                       # SUPPORTED: no MaD row or regular CodeQL model that I can find... (Note: check for existing model again)
+
+      # MAYBE Positive models (not sure about these):
+      - ["java.lang", "String", False, "valueOf", "(Object)", "", "Argument[0]", "ReturnValue", "taint", "manual"]                                               # **UNSUPPORTED**: already modelled in java.lang.model.yml at lines 82-84 for other signatures, this exact signature is not modelled, is there a reason to exclude it?
```
This is an alias for Object.toString, except the dispatch is hidden. That is to say, that this is a complex case that we haven't fully decided how to deal with.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+
+      # MAYBE Positive models (not sure about these):
+      - ["java.lang", "String", False, "valueOf", "(Object)", "", "Argument[0]", "ReturnValue", "taint", "manual"]                                               # **UNSUPPORTED**: already modelled in java.lang.model.yml at lines 82-84 for other signatures, this exact signature is not modelled, is there a reason to exclude it?
+      - ["java.lang", "String", False, "valueOf", "(int)", "", "Argument[0]", "ReturnValue", "taint", "manual"]                                                  # **UNSUPPORTED**: already modelled in java.lang.model.yml at lines 82-84 for other signatures, this exact signature is not modelled, is there a reason to exclude it?
```
Numbers are not taint-carriers by default.

aschackmull replied:
I guess we could have it, as most queries that want to ignore numbers use `int`-typed nodes as sanitizers anyway. But it' not important to include.

atorralba replied:
I'd prefer having this kind of models, since a few queries do care about tainted numbers. As you said, most queries add sanitizers for number and primitive types so it shouldn't have a negative impact.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      # MAYBE Positive models (not sure about these):
+      - ["java.lang", "String", False, "valueOf", "(Object)", "", "Argument[0]", "ReturnValue", "taint", "manual"]                                               # **UNSUPPORTED**: already modelled in java.lang.model.yml at lines 82-84 for other signatures, this exact signature is not modelled, is there a reason to exclude it?
+      - ["java.lang", "String", False, "valueOf", "(int)", "", "Argument[0]", "ReturnValue", "taint", "manual"]                                                  # **UNSUPPORTED**: already modelled in java.lang.model.yml at lines 82-84 for other signatures, this exact signature is not modelled, is there a reason to exclude it?
+      - ["java.lang", "String", False, "charAt", "(int)", "", "Argument[-1]", "ReturnValue", "taint", "manual"]                                                  # **UNSUPPORTED**: flow from String to returned char value?, is only getting a char back not considered interesting flow?
```
Similar reasoning, although this case is more tricky. We may want this as a general model, and we've discussed it before without a clear-cut answer, I believe.

jcogs33 replied:
Okay, thanks, I'll leave it in the "positive" section for now.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.lang", "RuntimeException", False, "RuntimeException", "(Throwable)", "", "Argument[0]", "Argument[-1]", "value", "manual"]                        # **UNSUPPORTED**: flow from arg to new RuntimeException object?
+      - ["java.lang", "IllegalArgumentException", False, "IllegalArgumentException", "(String)", "", "Argument[0]", "Argument[-1]", "value", "manual"]           # **UNSUPPORTED**: flow from arg to new IllegalArgumentException object?
+      - ["java.lang", "IllegalStateException", False, "IllegalStateException", "(String)", "", "Argument[0]", "Argument[-1]", "value", "manual"]                 # **UNSUPPORTED**: flow from arg to new IllegalStateException object?
+      - ["java.lang", "UnsupportedOperationException", False, "UnsupportedOperationException", "(String)", "", "Argument[0]", "Argument[-1]", "value", "manual"] # **UNSUPPORTED**: flow from arg to new UnsupportedOperationException object?
```
We're lacking some models for exceptions, but that's also because the support for exception flow hasn't been merged yet (it's blocked on performance issues). These models don't look right, though, they should probably use some more detailed field flow.

jcogs33 replied:
Thanks, I'll take a closer look at how to model these properly.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.lang", "String", "isEmpty", "()", "manual"]                # UNSUPPORTED: type as sanitizer
+      - ["java.lang", "String", "contains", "(CharSequence)", "manual"]   # UNSUPPORTED: type as sanitizer
+      - ["java.lang", "String", "startsWith", "(String)", "manual"]       # UNSUPPORTED: type as sanitizer
+      - ["java.lang", "StringBuilder", "toString", "()", "manual"]        # **SUPPORTED**: shouldn't model due to causing problems with dataflow?, unless it's fine with StringBuilder? (Note: modelled in java.lang.model.yml at line 34 through AbstractStringBuilder.toString subtyping)
```
This should not be in the negative models section, right? As you already note, it's already modelled.

jcogs33 replied:
@yo-h had mentioned that `toString` should not be modeled because it causes problems with dataflow, which is why I put it with the "negative" models. Is `StringBuilder.toString` safe to model? Or should its existing model be removed and it turned into a negative model?

atorralba replied:
`toString` in general, and particularly `Object.toString`, are unwanted models because they produce a lot of FP flow. But there are some classes which override `toString` in such a way that we're sure that they propagate taint, and `StringBuilder` is one of them. So we should keep the model and remove this row from the negative models.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.util", "Objects", "hash", "(Object[])", "manual"]          # UNSUPPORTED: do we care about controlling the hashcode value?
+      - ["java.lang", "Object", "hashCode", "()", "manual"]               # UNSUPPORTED: do we care about controlling the hashcode value?
+      - ["java.lang", "String", "hashCode", "()", "manual"]               # UNSUPPORTED: do we care about controlling the hashcode value?
+      - ["java.lang", "Throwable", "printStackTrace", "()", "manual"]     # UNSUPPORTED: only print stacktrace, no flow? (or flow from Throwable to its printed stacktrace??)
```
This should probably not be a general step, but there might be specialised queries that care.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.lang", "Enum", "name", "()", "", "manual"]                 # UNSUPPORTED: just gets name, no flow?, similar usage to toString() per Javadocs
+
+      # MAYBE Negative models (not sure about these):
+      - ["java.util", "Objects", "hash", "(Object[])", "manual"]          # UNSUPPORTED: do we care about controlling the hashcode value?
```
> do we care about controlling the hashcode value?
No

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.lang", "Object", "hashCode", "()", "manual"]               # UNSUPPORTED: do we care about controlling the hashcode value?
+      - ["java.lang", "String", "hashCode", "()", "manual"]               # UNSUPPORTED: do we care about controlling the hashcode value?
+      - ["java.lang", "Throwable", "printStackTrace", "()", "manual"]     # UNSUPPORTED: only print stacktrace, no flow? (or flow from Throwable to its printed stacktrace??)
+      - ["java.lang", "Throwable", "getMessage", "()", "", "manual"]      # UNSUPPORTED: only get message, no flow? (or flow from Throwable to its message??)
```
This should likely have a model, but as mentioned above, it's not really relevant yet.

aschackmull commented on java/ql/lib/ext/TopJdkApis.model.yml:
```diff
@@ -0,0 +1,133 @@
+      - ["java.lang", "String", "hashCode", "()", "manual"]               # UNSUPPORTED: do we care about controlling the hashcode value?
+      - ["java.lang", "Throwable", "printStackTrace", "()", "manual"]     # UNSUPPORTED: only print stacktrace, no flow? (or flow from Throwable to its printed stacktrace??)
+      - ["java.lang", "Throwable", "getMessage", "()", "", "manual"]      # UNSUPPORTED: only get message, no flow? (or flow from Throwable to its message??)
+      - ["java.sql", "ResultSet", "getString", "(String)", "", "manual"]  # UNSUPPORTED: similar to toString()?
```
No idea what this does. I'll note that this is the only `java.sql` method that made it onto the list. It looks like something that probably ought to have a model.

jcogs33 commented:
> We shouldn't duplicate rows, and I think these generally make more sense in the <packagename>.model.yml files. I believe going through this top100 was mainly meant to ensure that we do have manual models checked in where applicable, and that we hadn't missed any among the top 100.

Would it make sense to store these in a single file if maybe the model generator will be run on a list of top APIs like you suggested in the planning meeting today? Or at least tag the existing models somehow so they can be identified as top APIs by the generator?

yo-h commented:
> We shouldn't duplicate rows, and I think these generally make more sense in the <packagename>.model.yml files.

I think this makes sense - it will put the added data where it naturally belongs.

> Or at least tag the existing models somehow so they can be identified as top APIs by the generator?

Good point that we want to ideally retain the information about top APIs somewhere.

> I believe going through this top100 was mainly meant to ensure that we do have manual models checked in where applicable, and that we hadn't missed any among the top 100.

Perhaps capture the information about the list of top 100 APIs in a test case that checks that a manual model exists for each of the APIs in the list?

yo-h commented:
> Perhaps capture the information about the list of top 100 APIs in a test case that checks that a manual model exists for each of the APIs in the list?

This should be possible to do in a database-independent way. (Unlike running the metric queries or generator queries.)

aschackmull commented:
> Perhaps capture the information about the list of top 100 APIs in a test case that checks that a manual model exists for each of the APIs in the list?

Yes, I think that makes the most sense. The top 100 is a moving target (although unlikely to have changes that often), so taking a snapshot a wrapping that up as a test seems fine.

> This should be possible to do in a database-independent way. (Unlike running the metric queries or generator queries.)

Unfortunately, not quite. As an example we have `java.util.Set.add(Object)` in the top 100, but this is covered by a model for `Collection.add()`. However, as a qltest this can be easily fixed by adding a bit of Java to the Java side of the qltest to actually reference those 100 api endpoints, thus ensuring that they and their supertypes are in the db.

aschackmull commented on java/ql/lib/ext/java.lang.model.yml:
```diff
@@ -37,13 +37,23 @@ extensions:
       - ["java.lang", "CharSequence", True, "charAt", "", "", "Argument[-1]", "ReturnValue", "taint", "manual"]
       - ["java.lang", "CharSequence", True, "subSequence", "", "", "Argument[-1]", "ReturnValue", "taint", "manual"]
       - ["java.lang", "CharSequence", True, "toString", "", "", "Argument[-1]", "ReturnValue", "taint", "manual"]
+      - ["java.lang", "IllegalArgumentException", False, "IllegalArgumentException", "(String)", "", "Argument[0]", "Argument[-1].SyntheticField[java.lang.IllegalArgumentException.message]", "value", "manual"]
+      - ["java.lang", "IllegalArgumentException", False, "getMessage", "()", "", "Argument[-1].SyntheticField[java.lang.IllegalArgumentException.message]", "ReturnValue", "value", "manual"]
```
All of these `getMessage()` methods are overrides of the one on `Throwable`. Note that we would like the following to have flow:
```
Exception e = new RuntimeException("source");
sink(e.getMessage());
```
so we need all of these to use the same synthetic field - which makes sense, since the actual implementation uses just one message field on Throwable. That means that we can have just one model for `getMessage` on Throwable, which should then also apply to overrides.

You may want to add a test along the lines of the above - then you can see that it currently fails and check that it succeeds when you have updated the models.

jcogs33 replied:
Thanks! Updated in https://github.com/github/codeql/pull/11572/commits/08546549bf8585ebcf87899bbd2d140315c8f154.

aschackmull commented on java/ql/lib/ext/java.util.stream.model.yml:
```diff
@@ -87,3 +87,9 @@ extensions:
+      pack: codeql/java-all
+      extensible: extNeutralModel
+    data:
+      - ["java.util.stream", "Collectors", "toList", "()", "manual"]
```
You've excluded parameterless constructors from model consideration - you could do the same for parameterless static methods. And once we have that, I guess we should probably just drop this neutral model?

jcogs33 replied:
I've opened a PR for this: https://github.com/github/codeql/pull/11717
I believe the neutral models for the below APIs will need to be dropped as well:
- `java.lang.System#currentTimeMillis()`
- `java.util.Collections#emptyList()`
- `java.util.Optional#empty()`
- `java.util.UUID#randomUUID()`

aschackmull commented on java/ql/test/ext/TopJdkApis/TopJdkApisTest.expected:
```diff
@@ -0,0 +1,97 @@
+| java.util.stream.Stream#filter(Predicate) |
+| java.util.stream.Stream#findFirst() |
+| java.util.stream.Stream#forEach(Consumer) |
+| java.util.stream.Stream#map(Function) |
```
There are 97 lines here. Was this meant to include all 100?

jcogs33 replied:
It should be 97 instead of 100 due to the fact that [`java.util.stream.Stream#collect(Collector)`](https://github.com/github/codeql/pull/11572#discussion_r1040967124), [`java.lang.String#valueOf(Object)`](https://github.com/github/codeql/pull/11572#discussion_r1040978328), and [`java.lang.Throwable#printStackTrace()`](https://github.com/github/codeql/pull/11572#discussion_r1040991878) are not modeled. 
Let me know if I should add a neutral model for any of these instead of leaving them unmodeled.

aschackmull replied:
No, I think that's fine as-is. Once the test is restructured as I've suggested, then it should be more clear anyway, as these will be the only results.

aschackmull commented on java/ql/test/ext/TopJdkApis/TopJdkApis.qll:
```diff
@@ -0,0 +1,90 @@
+          "#" + this.getName() + paramsString(this) and
+      apiName in [
+          // top 100 JDK APIs
+          "java.lang.StringBuilder#append(String)", "java.util.List#get(int)",
```
I think it would be good to pull this string list into its own predicate `topJdkApiName(string apiname)`. If you then also expose the mapping between apiname and callable, then the test could report 1. those apinames for which there's no callable (i.e. an import is missing in the .java file), and 2. those apinames for which there _isn't_ a manual model. That way we could have an empty expected file and more confidence that all the relevant imports are present.

jcogs33 replied:
Updated in https://github.com/github/codeql/pull/11572/commits/fdcaa93200c1e422b4c3239dd34e0b1f80fd31b6.
Let me know if I misunderstood anything about what you had in mind for this.

aschackmull commented on java/ql/test/ext/TopJdkApis/TopJdkApis.qll:
```diff
@@ -0,0 +1,103 @@
+}
+
+predicate hasCallable(string apiName) {
+  exists(Callable c |
```
If you make this a parameter of the predicate, then you can reduce some code duplication, as the body of this predicate is repeated 3 times (here, in the definition of TopJdkApi, and in the test).

jcogs33 replied:
I don't think I'm fully understanding the code structure you have in mind for this. 🤔 
I've added a `string getApiName(Callable api)` predicate in https://github.com/github/codeql/pull/11572/commits/c0628035fadd40774786deb28ee22bc574db4cb4 for now to somewhat reduce the duplication. Let me know if that structure is alright, otherwise I'll think about this some more to see if I can come up with what you are proposing.

aschackmull commented on java/ql/test/ext/TopJdkApis/TopJdkApisTest.ql:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @name Test for top JDK APIs
+ * @description Provides a test case to check that a manual models exists for each each of the top JDK APIs.
+ * @id java/top-jdk-apis-test
+ */
```
A qltest does not need metadata, so this block can be deleted.

aschackmull commented on java/ql/test/ext/TopJdkApis/TopJdkApisTest.ql:
```diff
@@ -0,0 +1,23 @@
+from string apiName, TopJdkApi topApi
+where
+  // top jdk api names for which there is no callable
+  topJdkApiName(apiName) and
+  not hasCallable(apiName)
+  or
+  // top jdk api names for which there isn't a manual model
+  not topApi.hasManualMadModel() and
+  apiName =
+    topApi.asCallable().getDeclaringType().getPackage() + "." +
+      topApi.asCallable().getDeclaringType().getSourceDeclaration() + "#" +
+      topApi.asCallable().getName() + paramsString(topApi.asCallable())
```
This should be fixed.

aschackmull replied:
I.e. move the variable declaration into the disjunct as an `exists`.

aschackmull commented on java/ql/test/ext/TopJdkApis/TopJdkApisTest.ql:
```diff
@@ -0,0 +1,23 @@
+    topApi.asCallable().getDeclaringType().getPackage() + "." +
+      topApi.asCallable().getDeclaringType().getSourceDeclaration() + "#" +
+      topApi.asCallable().getName() + paramsString(topApi.asCallable())
+select apiName order by apiName
```
Add a message column that states which of the two cases we're in, so we can see in the test output whether there's no callable, or whether there isn't a manual model.

jcogs33 replied:
Updated in https://github.com/github/codeql/pull/11572/commits/640b450c4783aed2061b58fcb850d79cd414b40c.

aschackmull commented on java/ql/lib/ext/java.lang.model.yml:
```diff
@@ -37,13 +37,19 @@ extensions:
       - ["java.lang", "CharSequence", True, "charAt", "", "", "Argument[-1]", "ReturnValue", "taint", "manual"]
       - ["java.lang", "CharSequence", True, "subSequence", "", "", "Argument[-1]", "ReturnValue", "taint", "manual"]
       - ["java.lang", "CharSequence", True, "toString", "", "", "Argument[-1]", "ReturnValue", "taint", "manual"]
+      - ["java.lang", "IllegalArgumentException", False, "IllegalArgumentException", "(String)", "", "Argument[0]", "Argument[-1].SyntheticField[java.lang.Throwable.message]", "value", "manual"]
```
It would probably be nice to include a few more typical exception constructor models. Just including the subset that falls within the current top100 ends up being a bit arbitrary. There are IIRC both `(String)` and `(String,Throwable)` variations on most (all?) of them, and we very likely also want to include `Exception` and `IOException`.

jcogs33 replied:
Can I include these on a separate PR since this PR was supposed to be for the top-100?

aschackmull replied:
sure

jcogs33 commented:
**Review Consideration** for commit https://github.com/github/codeql/pull/11572/commits/bf6148c47705a2285d85b9949e55cafbf15df65b onwards:

1.  Is the code style for [`TopJdkApis.qll`](https://github.com/github/codeql/pull/11572/files#diff-99459acbe9ddfc84ab9e7d70b9deb793ab8c9216eca9d9cd987ec68c748855b1) and [`TopJdkApisTest.ql`](https://github.com/github/codeql/pull/11572/files#diff-1bf26f7a77313f7fdb94e233eda61fec70a278bd9d824d882471b003ee6ca396) okay? Should anything else be updated? See [#11572 comment](https://github.com/github/codeql/pull/11572#discussion_r1050578584).
2.  In https://github.com/github/codeql/pull/11572/commits/f3fc68352eff3e3fb9673050c1a96cf45dfd7a79 and https://github.com/github/codeql/pull/11572/commits/f933fc75cde922aac959acbb69de1d049ccae4ab, I updated some existing test cases that were affected by the addition of the `Integer.parseInt` MaD model. I'm a little bit confused as to why these needed to be updated since it looks like `Integer.parseInt` was already supported by [`NumberTaintPreservingCallable`](https://github.com/github/codeql/blob/30451ee95088bca2707c7d8e9da2d01066d1d59e/java/ql/lib/semmle/code/java/dataflow/FlowSteps.qll#L118). Please let me know if anything looks wrong with the MaD model.
3.  Regarding [this comment](https://github.com/github/codeql/pull/11572/files#r1049415888), would it be okay to leave the parameterless static methods as neutral models for now and remove them later if an exclusion specific to summary models is added? See [#11717 comment](https://github.com/github/codeql/pull/11717#issuecomment-1354772131).
4.  Let me know if the category and description in the [change note](https://github.com/github/codeql/pull/11572/files#diff-8d892e9bf034438de544869b3c53c0bf22275b7cb1280fd7fa5294624c5a2f3c) look okay.

atorralba commented:
> Is the code style for [TopJdkApis.qll](https://github.com/github/codeql/pull/11572/files#diff-99459acbe9ddfc84ab9e7d70b9deb793ab8c9216eca9d9cd987ec68c748855b1) and [TopJdkApisTest.ql](https://github.com/github/codeql/pull/11572/files#diff-1bf26f7a77313f7fdb94e233eda61fec70a278bd9d824d882471b003ee6ca396) okay? Should anything else be updated? See https://github.com/github/codeql/pull/11572#discussion_r1050578584.

I _think_ what you did is pretty similar to what @aschackmull meant, with the addition that you could have `Callable c` as a parameter of `hasCallable` (renamed to e.g. `hasApiName`), making `getApiName` unnecessary. So something like:

```ql
/** Holds if `c` has the MaD-formatted name `apiName`. */
predicate hasApiName(Callable c, string apiName) {
  apiName =
    c.getDeclaringType().getPackage() + "." + c.getDeclaringType().getSourceDeclaration() + "#" +
      c.getName() + paramsString(c)
}
```

Which would still remove the code duplication, since the uses would look like:

```ql
class TopJdkApi extends SummarizedCallableBase {
  TopJdkApi() {
    exists(string apiName |
      hasApiName(this.asCallable(), apiName) and
      topJdkApiName(apiName)
    )
  }
}
```

and

```ql
 // top jdk api names for which there isn't a manual model
 exists(TopJdkApi topApi |
  not topApi.hasManualMadModel() and
  hasApiName(topApi.asCallable(), apiName) and
  message = "no manual model"
)
```

> In https://github.com/github/codeql/commit/f3fc68352eff3e3fb9673050c1a96cf45dfd7a79 and https://github.com/github/codeql/commit/f933fc75cde922aac959acbb69de1d049ccae4ab, I updated some existing test cases that were affected by the addition of the Integer.parseInt MaD model. I'm a little bit confused as to why these needed to be updated since it looks like Integer.parseInt was already supported by [NumberTaintPreservingCallable](https://github.com/github/codeql/blob/30451ee95088bca2707c7d8e9da2d01066d1d59e/java/ql/lib/semmle/code/java/dataflow/FlowSteps.qll#L118). Please let me know if anything looks wrong with the MaD model.

Those changes only add new edges and nodes, but the select results are the same. I'm not an expert, but AFAIK that has to do with how MaD rows are represented in the dataflow graph compared to regular CodeQL steps. So nothing you should worry about as long as the results don't change.

> Regarding [this comment](https://github.com/github/codeql/pull/11572/files#r1049415888), would it be okay to leave the parameterless static methods as neutral models for now and remove them later if an exclusion specific to summary models is added? See https://github.com/github/codeql/pull/11717#issuecomment-1354772131.

Given that it seems that the model generator will still be generally interested in parameterless static methods (just not as sinks nor summaries), I think it's a good idea to leave them as neutral models for the moment. If an hypothetical exclusion for summary models happens, those can be removed.

> Let me know if the category and description in the [change note](https://github.com/github/codeql/pull/11572/files#diff-8d892e9bf034438de544869b3c53c0bf22275b7cb1280fd7fa5294624c5a2f3c) look okay.

I think it looks right. As an optional nitpick, I'm not sure if specifying the kind of models (summary and neutral) is of general interest, so maybe a simpler `Added more dataflow models for top-100 JDK APIs.` would work.

I felt tempted to categorize this as a `majorAnalysis` change given the widespread impact that changes in JDK models could have, but I think `majorAnalysis` is more appropriate for significant CodeQL API changes and not just modeling changes, so `minorAnalysis` is probably the right category.

jcogs33 commented:
Thanks @atorralba! I've added a `hasApiName` predicate in https://github.com/github/codeql/pull/11572/commits/42ddd6636059566ec3a35385813ab32b2625acc3 and updated the change note as suggested in https://github.com/github/codeql/pull/11572/commits/f37f0a09aa1bf013430d21f2c6f16b6beddc75b0.

atorralba commented on java/ql/test/ext/TopJdkApis/TopJdkApis.qll:
```diff
@@ -0,0 +1,101 @@
+
+  /** Holds if this API has a manual summary model. */
+  private predicate hasManualSummary() {
+    this instanceof SummarizedCallable and
+    this.(SummarizedCallable).hasProvenance(false)
```
`this instanceof SummarizedCallable` is implied by the cast in the next line:

```suggestion
    this.(SummarizedCallable).hasProvenance(false)
```

atorralba commented on java/ql/test/ext/TopJdkApis/TopJdkApis.qll:
```diff
@@ -0,0 +1,101 @@
+
+  /** Holds if this API has a manual neutral model. */
+  private predicate hasManualNeutral() {
+    this instanceof FlowSummaryImpl::Public::NeutralCallable and
+    this.(FlowSummaryImpl::Public::NeutralCallable).hasProvenance(false)
```
Same thing here.

```suggestion
    this.(FlowSummaryImpl::Public::NeutralCallable).hasProvenance(false)
```

atorralba commented on java/ql/test/ext/TopJdkApis/TopJdkApisTest.ql:
```diff
@@ -0,0 +1,17 @@
+where
+  // top jdk api names for which there is no callable
+  topJdkApiName(apiName) and
+  not exists(Callable c | hasApiName(c, apiName)) and
```
Small simplification:

```suggestion
  not hasApiName(_, apiName) and
```

jcogs33 replied:
Thanks! Updated all in https://github.com/github/codeql/pull/11572/commits/a8ee633acd19861f94834effb24de72991f06414.

atorralba reviewed (commented):
Looks great! Just a few more minor suggestions after taking a closer look.

atorralba reviewed (dismissed):
Happy to proxy @aschackmull almost-complete review (I didn't re-review in full) and get this merged if DCA is happy 😃.

jcogs33 commented:
Thanks @atorralba! DCA looks good now I think (the only failures are the typical `apache_commons` ones). 
Can you re-approve when you have a chance? :pray: I pushed one very minor update to store the neutral models alphabetically.
